templates/fine-tune-function-calling/fc_utils/eval_utils.py

```diff
@@ -0,0 +1,163 @@
+            return False, PossibleMistakes.WRONG_FUNCTION_NAME.value
+        elif expected_tool_call["arguments"] != actual_tool_call["arguments"]:
+            if len(expected_tool_call["arguments"]) != len(actual_tool_call["arguments"]):
+                return False, PossibleMistakes.MISSING_ARGUMENT.value
+            return False, PossibleMistakes.WRONG_ARGUMENT_VALUE.value
+    return True, ""
+
+
+def parse_and_eval(
+    parser: Union[OpenAIResponseParser, AnyscaleResponseParser],
+    messages,
+    expected_responses,
+    tools=None,
+):
+    """
+    Parse and eval loop to parse the assistant responses and evaluate them against the ground truth in the conversation.
+    Args:
+    parser: OpenAIResponseParser or AnyscaleResponseParser object
+    messages: list of messages in the conversation. The messages with role 'assistant' are ignored and triggers a chat completion call
+    expected_responses: list of ground truth responses
+    tools: list of tools to available, used by OpenAIResponseParser
+    """
+    assert messages[0]["role"] == "system", "First message must be from system"
+    match = True
+    current_conv = []
+    assist_id = 0
+    conv_id = 0
+    while conv_id < len(messages):
+        if messages[conv_id]["role"] != "assistant":
+            current_conv.append(messages[conv_id])
+        # 'role' is 'assistant'. Thus, get assistant response
+        else:
+            if current_conv[-1]["role"] == "tool":
+                # If the last message was a tool, we need additional processing.
+                # For a tool, we need to have a valid tool call id. This can only be retrieved from the previous assistant response
+                # because our dataset is synthetically constructed
+                if (
+                    current_conv[-2]["role"] == "assistant"
+                    and current_conv[-2]["tool_calls"]
+                ):
+                    current_conv[-1]["tool_call_id"] = current_conv[-2]["tool_calls"][0].id
+                else:
+                    return current_conv, False
+            if isinstance(parser, OpenAIResponseParser):
+                parsed_response = parser.get_parsed_response(current_conv, tools)
+            else:
+                parsed_response = parser.get_parsed_response(current_conv)
+            if parsed_response["content"] == ERROR_OUTPUT:
+                return None, None  # return None if there's an error
+
+            is_correct, mistake_type = is_match(parsed_response, expected_responses[assist_id])
+            match = match and (is_correct)
+            original_assistant_response = parsed_response["original_response"]
+            current_conv.append(dict(original_assistant_response))
+            assist_id += 1 # next expected response
+            if not match:  # return right away if model output is incorrect
+                return current_conv, match, mistake_type
+        conv_id += 1 # next message
+    return current_conv, match, mistake_type
```

Comment:
here return is three items but above you are returning two items. Please fix to be consistent.

Reply:
Whoa. Fixed.